haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs

```diff
@@ -57,30 +71,100 @@ withTypechecked globalLock GhcArgs{..} analysisOpts xrefSink
         = withMVar globalLock . const . errHandling $ do
     -- TODO(robinpalotai): logging
     printErr "Running GHC"
-    -- Make sure we don't use any tools except cpp.
-    let modifyFlags = maybe id setPgmP (overridePgmP gaToolOverride)
-                    . dontGenerateCode
-                    . verbose 0
     xrefGraph <- runGhc (Just $ gaLibdirPrefix </> libdir) $ do
         -- see GHC trac #4162
         liftIO . void $ installHandler sigINT Default Nothing
+        -- The obscure preparation that follows is explained in
+        -- https://gist.github.com/robinp/49c68c6f69f6aabfddee3cba42b4964f,
+        -- further referenced below as 'the gist'.
         dflags0 <- getSessionDynFlags
-        (parsedFlags, unused, errors) <- parseDynamicFlags dflags0
-                                            (map noLoc gaArgs)
-        printErr $ "Yet unused flags (target candidates): "
+        let defaultHscTarget = hscTarget dflags0
+            defaultLink = ghcLink dflags0
+            args0 = gaArgs
+            -- This is our baseline config, to be modified by actual flags.
+            dflags1 = (dontGenerateCode . verbose 0) dflags0
```

Comment:
dflags1, 2 and 3 have semantic meaning. Maybe rename them to preserve that?
Readers might miss the connection between the number and the above git gist. Maybe a simple comment what 1, 2 and 3 refer to will be sufficient?

Reply:
Not sure - with the numbering it's easier to spot if a non-recent dflag would be passed as argument during a refactoring/change (and this code is a bit prone to change). The scope is small enough here to see the specific change from the comments/operations made.

A change I could imagine is to keep only a current dflags in an IORef, to avoid accidental confusion. WDYT?

Reply:
Keeping only the current dflags in IORef sounds good to me.

Reply:
I don't understand all of these changes but using an `IORef` to store a mutable value seems like the wrong solution. I think for this code being very explicit about which functions are updating the `DynFlags` is very important when tracking down bugs. If instead you have random functions mutating state, it is much harder to find the points where the `DynFlags` are modified.

Reply:
I went with a third solution: keeping the DynFlags in the Ghc monad's session. One would need to put the DynFlags into the session anyway for them to be used, so it's a convenient mutable place.

I also added some helpers that make it more explicit where flag modification / read happens. The flag mutating code is pretty self-contained here (a few screens), certainly not spreading out. I tried to go with the immutable way, but I found myself setting the flags on the session repeatedly, also refactoring was hard since I needed to add `dflags3b` and make sure that the calculation of `dflags4` at all places depended on `3b` and not `3`, which seems a lot more prone to errors.

In case an indexing failure happens, it should still be easy to reproduce and see what goes wrong.